include/dali/core/mm/detail/free_list.h

```diff
@@ -0,0 +1,327 @@
+ * auxiliary list of unused block descriptors. These blocks will be reused when something
+ * is added to the list to avoid interaction with the heap.
+ */
+class uniform_free_list {
+ public:
+  struct block {
+    block *next;
+    void *mem;
+  };
+
+  uniform_free_list() = default;
+  uniform_free_list(const uniform_free_list &) = delete;
+  uniform_free_list(uniform_free_list &&other) {
+    swap(other);
+  }
+
+  void swap(uniform_free_list &other) {
+    std::swap(head_, other.head_);
+    std::swap(unused_blocks_, other.unused_blocks_);
+  }
+
+  uniform_free_list &operator=(uniform_free_list &&other) {
+    if (this != &other) {
+      clear();
+      swap(other);
+    }
+    return *this;
+  }
+
+  ~uniform_free_list() {
+    clear();
+  }
+
+  void clear() {
+    free_list_impl::delete_list(head_);
+    free_list_impl::delete_list(unused_blocks_);
+  }
+
+  void *get() {
+    if (block *blk = head_) {
+      void *mem = blk->mem;
+      head_ = blk->next;
+      blk->next = unused_blocks_;
+      blk->mem  = nullptr;
+      unused_blocks_ = blk;
+      return mem;
+    } else {
+      return nullptr;
+    }
+  }
+
+  void put(void *ptr) {
+    block *blk = unused_blocks_;
+    if (blk) {
+      unused_blocks_ = blk->next;
+      blk->next = head_;
+      blk->mem = ptr;
+    } else {
+      blk = new block{head_, ptr};
+    }
+    head_ = blk;
+  }
+
+ private:
+  block *head_ = nullptr;
+  block *unused_blocks_ = nullptr;
+};
+
+/**
+ * @brief Maintains a list of free memory blocks of variable size, returning free blocks
+ *        with least margin.
+ *
+ * This object is used for managing a list of variable-sized memory regions.
+ * It returns blocks on a best-fit basis. The remaining parts of the block are placed in the list.
+ *
+ * This list may be of utility when there's a limited range of object sizes to be allocated
+ * - ideally, if these are powers of two. It suffers from fragmentation and therefore is not
+ * intended for long term usage.
+ */
+class best_fit_free_list {
+ public:
+  best_fit_free_list() = default;
+  best_fit_free_list(const best_fit_free_list &) = delete;
+  best_fit_free_list(best_fit_free_list &&other) {
+    swap(other);
+  }
+
+  void swap(best_fit_free_list &other) {
+    std::swap(head_, other.head_);
+    std::swap(unused_blocks_, other.unused_blocks_);
+  }
+
+  best_fit_free_list &operator=(best_fit_free_list &&other) {
+    if (this != &other) {
+      clear();
+      swap(other);
+    }
+    return *this;
+  }
+
+  virtual ~best_fit_free_list() {
+    clear();
+  }
+
+  void clear() {
+    free_list_impl::delete_list(head_);
+    free_list_impl::delete_list(unused_blocks_);
+  }
+
+  struct block {
+    block *next;
+    char  *start, *end;
+
+    size_t fit(size_t requested, size_t alignment) const {
+      char *base = static_cast<char *>(start);
+      char *aligned = detail::align_ptr(base, alignment);
+      return (end - aligned) - requested;  // intentional wrap into huge values
```

Comment:
Can it wrap to something smaller than `(static_cast<size_t>(-1)) >> 1;` ? so L184 will find it as a fit?

Reply:
Not with sane values (on 64-bit systems).